### PR TITLE
Wrap Monolog handlers to handle exceptions gracefully

### PR DESCRIPTION
**Pull request by nathanielks**
A client observed elevated error rates when connections to the fluent bit container timed out. This wraps the log handlers to catch any exceptions thrown by the handler and log them gracefully.

### CODE REVIEW

File: inc/namespace.php [hunk 1]
```diff
@@ -1019,7 +1019,11 @@ function get_logger( string $log_group, string $log_stream ) : LoggerInterface {
 		// otherwise it cannot parse the log entries.
 		$socket->setFormatter( new MsgPackFormatter() );
 
-		$logger->pushHandler( $socket );
+		// Catches any exceptions thrown by Monolog itself and logs them via
+		// error_log
```
Comment (roborourke) [inline inc/namespace.php:1023]:
```suggestion
		// Catches any exceptions thrown by Monolog itself and logs them via error_log().
```

File: inc/class-loggerexceptionhandler.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+/*
+ * Catches exceptions thrown by Monolog and reports them using error_log.
+ *
+ */
+
+use Monolog\Handler\HandlerWrapper;
+use Throwable;
+
+class LoggerExceptionHandler extends HandlerWrapper {
```
Comment (roborourke) [inline inc/class-loggerexceptionhandler.php:18]:
```suggestion
/**
 * Catches exceptions thrown by Monolog and reports them using error_log.
 */
class LoggerExceptionHandler extends HandlerWrapper {
```

File: inc/class-loggerexceptionhandler.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+	/**
```
Comment (roborourke) [inline inc/class-loggerexceptionhandler.php:20]:
```suggestion
	/**
	 * Check handler for exceptions.
	 *
```

File: inc/class-loggerexceptionhandler.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+			return $this->handler->handle( $record );
+		} catch ( Throwable $e ) {
+			$this->logException( $e );
+		}
+
+		return true;
+	}
+
+	/**
+	 * {@inheritdoc}
```
Comment (roborourke) [inline inc/class-loggerexceptionhandler.php:34]:
```suggestion
	 * Catch batch exceptions.
	 *
	 * {@inheritdoc}
```

File: inc/namespace.php [hunk 5]
```diff
@@ -1042,7 +1046,11 @@ function get_logger( string $log_group, string $log_stream ) : LoggerInterface {
 		$formatter = new LineFormatter( '%message%' );
 		$handler->setFormatter( $formatter );
 
-		$logger->pushHandler( $handler );
+		// Catches any exceptions thrown by Monolog itself and logs them via
+		// error_log
```
Comment (roborourke) [inline inc/namespace.php:1050]:
```suggestion
		// Catches any exceptions thrown by Monolog itself and logs them via error_log.
```

File: inc/class-loggerexceptionhandler.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
+<?php
+/**
+ * Altis Cloud Fluent Bit MsgPack formatter.
```
Comment (roborourke) [inline inc/class-loggerexceptionhandler.php:3]:
```suggestion
 * Altis Cloud Monolog Exception Handler.
```

File: inc/class-loggerexceptionhandler.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+ * @package altis/cloud
+ */
+
+namespace Altis\Cloud;
+
+/*
+ * Catches exceptions thrown by Monolog and reports them using error_log.
+ *
+ */
+
```
Comment (roborourke) [inline inc/class-loggerexceptionhandler.php:14]:
```suggestion
```

**Review by roborourke — COMMENTED** [review]
Would calling the standard `error_log()` not still be routed to FluentBit then? I'm not sure how this works exactly if we're overriding the default log handler.

File: inc/class-loggerexceptionhandler.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Catches exceptions thrown by Monolog and reports them using error_log.
+ */
+class LoggerExceptionHandler extends HandlerWrapper {
+
+	/**
+	 * Check handler for exceptions.
+	 *
+	 * {@inheritdoc}
```
Comment (roborourke) [inline inc/class-loggerexceptionhandler.php:21]:
```suggestion
	 * {@inheritDoc}
```

**Comment by nathanielks** [comment]
> Would calling the standard error_log() not still be routed to FluentBit then? I'm not sure how this works exactly if we're overriding the default log handler.

Nope! `error_log()` isn't modified by Monolog (and can't be, PHP doesn't let anything change the behavior of `error_log()` besides where it routes logs). The only logs that Monolog is responsible for is those sent via `Altis\Cloud\get_logger('foo', 'bar')->info('message');`

**Comment by nathanielks** [comment]
Thank you for the linting fixes, btw!

**Comment by roborourke** [comment]
Yeah - I was conflating that and `trigger_error()`. Thanks for clarifying!

File: inc/class-loggerexceptionhandler.php [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Catches exceptions thrown by Monolog and reports them using error_log.
+ */
+class LoggerExceptionHandler extends HandlerWrapper {
+
+	/**
+	 * Check handler for exceptions.
+	 *
+	 * {@inheritDoc}
```
Comment (roborourke) [inline inc/class-loggerexceptionhandler.php:21]:
```suggestion
	 * @param array $record The structured error log record.
	 * @return bool
```

File: inc/class-loggerexceptionhandler.php [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+			$this->logException( $e );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Catch batch exceptions.
+	 *
+	 * {@inheritdoc}
```
Comment (roborourke) [inline inc/class-loggerexceptionhandler.php:36]:
```suggestion
	 * @param array $records The batch of structured error log records.
	 * @return void
```

**Review by roborourke — APPROVED** [review]

**Pull request merged**